v3_usm.go

```diff
@@ -622,28 +622,89 @@ func (sp *UsmSecurityParameters) discoveryRequired() *SnmpPacket {
 }
 
 func (sp *UsmSecurityParameters) calcPacketDigest(packet []byte) []byte {
-	var mac hash.Hash
+	return calcPacketDigest(packet, sp)
+}
 
-	switch sp.AuthenticationProtocol {
-	default:
-		mac = hmac.New(crypto.MD5.New, sp.SecretKey)
-	case SHA:
-		mac = hmac.New(crypto.SHA1.New, sp.SecretKey)
-	case SHA224:
-		mac = hmac.New(crypto.SHA224.New, sp.SecretKey)
-	case SHA256:
-		mac = hmac.New(crypto.SHA256.New, sp.SecretKey)
-	case SHA384:
-		mac = hmac.New(crypto.SHA384.New, sp.SecretKey)
-	case SHA512:
-		mac = hmac.New(crypto.SHA512.New, sp.SecretKey)
-	}
+// calcPacketDigest calculate authenticate digest for incoming messages (TRAP or
+// INFORM).
+// Support MD5, SHA1, SHA224, SHA256, SHA384, SHA512 protocols
+func calcPacketDigest(packetBytes []byte, secParams *UsmSecurityParameters) []byte {
+	var digest []byte
+
+	switch secParams.AuthenticationProtocol {
+	case MD5, SHA:
+		digest = digestRFC3414(
+			secParams.AuthenticationProtocol,
+			packetBytes,
+			secParams.SecretKey)
+	case SHA224, SHA256, SHA384, SHA512:
+		digest = digestRFC7860(
+			secParams.AuthenticationProtocol,
+			packetBytes,
+			secParams.SecretKey)
+	}
+
+	return digest
+}
 
+// digestRFC7860 calculate digest for incoming messages using HMAC-SHA2 protcols
+// according to RFC7860 4.2.2
+func digestRFC7860(h SnmpV3AuthProtocol, packet []byte, authKey []byte) []byte {
+	mac := hmac.New(h.HashType().New, authKey)
 	_, _ = mac.Write(packet)
 	msgDigest := mac.Sum(nil)
 	return msgDigest
 }
 
+// digestRFC3414 calculate digest for incoming messages using MD5 or SHA1
+// according to RFC3414 6.3.2 and 7.3.2
+func digestRFC3414(h SnmpV3AuthProtocol, packet []byte, authKey []byte) []byte {
+	var extkey [64]byte
+	var err error
+	var k1, k2 [64]byte
+	var h1, h2 hash.Hash
+
+	copy(extkey[:], authKey)
+
+	switch h {
+	case MD5:
+		h1 = md5.New()
+		h2 = md5.New()
+	case SHA:
+		h1 = sha1.New()
+		h2 = sha1.New()
```

Comment:
Could you change this to
```
               h1 = sha1.New() //nolint:gosec
               h2 = sha1.New() //nolint:gosec
```

Reply:
Thanks. I was strugling with linting.

By the way, linter says : a blank import should be only in a main or test package, or have a comment justifying it
I will be more than happy to change that, but this wasn't my doing.
I'm not really sure about it.

Reply:
> Thanks. I was strugling with linting.

No problem. For future reference, most of the linting checks can be done locally by running:
``docker build -t gosnmp/gosnmp:latest . && docker run -it gosnmp/gosnmp:latest``

> By the way, linter says : a blank import should be only in a main or test package, or have a comment justifying it
> I will be more than happy to change that, but this wasn't my doing.
> I'm not really sure about it.

Need to check up on this one.

Reply:
We need to keep these blank imports in order to register sha2 protocols.
I comment the blank import yo be useful.

Alternative would be to change code to make use of crypto/<sha2-protocols> and get ride of blank import.